org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend

```diff
@@ -1982,6 +2012,71 @@ class DocumentRangeFormattingParams extends DocumentFormattingParams {
 	}
 }
 
+/**
+ * The type hierarchy request is sent from the client resolve a {@link TypeHierarchyItem type hierarchy item} for
+ * a give cursor location in the text document. The request would also allow to specify if the item should be resolved
+ * and whether sub- or supertypes are to be resolved.
+ */
+@JsonRpcData
+class TypeHierarchyParams extends TextDocumentPositionParams {
+
+	/**
+	 * The number of hierarchy levels to resolve. {@code 0} indicates no hierarchy level. It defaults to {@code 0}.
+	 */
+	int resolve
+	
+	/**
+	 * The direction of the type hierarchy resolution. If not defined, defaults to {@code children}.
+	 * 
+	 * <p>
+	 * The followings are the legal values:
+	 * <ul>
+	 * <li>{@code children},</li>
+	 * <li>{@code parents}, and</li>
+	 * <li>{@code both}.</li>
+	 * </ul>
+	 * </p>
+	 */
+	String direction
+}
+
+/**
+ * Request to resolve an unresolved {@link TypeHierarchyItem type hierarchy item} which is indicated if the
+ * {@link TypeHierarchyItem#getParents parents} or the {@link TypeHierarchyItem#getChildren children} is not
+ * defined. If resolved and no {@code parents} or {@code children} are available then an empty list is returned.
+ */
+@JsonRpcData
+class ResolveTypeHierarchyItemParams {
+
+	/**
+	 * The hierarchy item to resolve.
+	 */
+	@NonNull
+	TypeHierarchyItem item
+
+	/**
+	 * The number of hierarchy levels to resolve. {@code 0} indicates no hierarchy level.
+	 */
+	@NonNull
+	Integer resolve
```

Comment:
We used `int` for other parts of the protocol that expect a non-nullable integer value.